crates/fuel-streams-publisher/src/telemetry/mod.rs

```diff
@@ -0,0 +1,248 @@
+mod elastic_search;
+mod publisher;
+mod system;
+
+use std::{sync::Arc, time::Duration};
+
+use anyhow::Context;
+use elastic_search::{
+    new_elastic_search,
+    should_use_elasticsearch,
+    ElasticSearch,
+    LogEntry,
+};
+use fuel_streams_core::prelude::*;
+// TODO: Consider using tokio's Rwlock instead
+use parking_lot::RwLock;
+use publisher::PublisherMetrics;
+use system::{System, SystemMetricsWrapper};
+use tokio::runtime::{Builder, Runtime};
+
+#[derive(Debug, Clone)]
+
+pub struct Telemetry {
+    runtime: Arc<Runtime>,
+    system: Arc<RwLock<System>>,
+    publisher_metrics: Arc<PublisherMetrics>,
+    elastic_search: Option<Arc<ElasticSearch>>,
+}
+
+impl Telemetry {
+    const DEDICATED_THREADS: usize = 2;
+
+    pub async fn new() -> anyhow::Result<Arc<Self>> {
+        let runtime = Builder::new_multi_thread()
+            .worker_threads(Self::DEDICATED_THREADS)
+            .enable_all()
+            .build()
+            .expect("Failed to create Telemetry runtime");
+
+        let system = Arc::new(RwLock::new(System::new().await));
+        let publisher_metrics = PublisherMetrics::new(None)
+            .expect("Publisher metrics must be created");
+
+        let elastic_search = if should_use_elasticsearch() {
+            Some(Arc::new(new_elastic_search().await?))
+        } else {
+            None
+        };
+
+        Ok(Arc::new(Self {
+            runtime: Arc::new(runtime),
+            system,
+            publisher_metrics: Arc::new(publisher_metrics),
+            elastic_search,
+        }))
+    }
+
+    pub async fn start(&self) -> anyhow::Result<()> {
+        let system = Arc::clone(&self.system);
+
+        if let Some(elastic_search) = self.elastic_search.as_ref() {
+            tracing::info!(
+                "Elastic Search connection live? {:?}",
+                elastic_search.get_conn().check_alive().unwrap_or_default()
+            );
+            elastic_search
+                .get_conn()
+                .ping()
+                .await
+                .context("Error pinging elastisearch connection")?;
+            tracing::info!("Elastic logger pinged successfully!");
+        };
+
+        self.runtime.spawn(async move {
+            System::monitor(system, Duration::from_secs(2)).await;
+        });
+
+        Ok(())
+    }
+
+    pub fn log_info(&self, message: &str) {
+        let entry = LogEntry::new("INFO", message);
+        self.maybe_elog(entry);
+        tracing::info!("{}", message);
+    }
+
+    pub fn log_error(&self, message: &str) {
+        let entry = LogEntry::new("ERROR", message);
+        self.maybe_elog(entry);
+        tracing::error!("{}", message);
+    }
+
+    fn maybe_elog(&self, entry: LogEntry) {
+        if let Some(elastic_search) = &self.elastic_search {
+            self.runtime
+                .spawn(elastic_search::log(elastic_search.clone(), entry));
+        }
+    }
```

Comment:
This happens within the Telemetry's dedicated ThreadPool, so, we no longer have to delay processing the next block because of ElasticSearch